### PR TITLE
Two simplifications that will be needed to deal with @zoep's issues regarding initcode equivalence

### PR DESCRIPTION
**Pull request by msooseth**
## Description
One more simplification of `readByte`. This one works ALSO when `CopySlice` is not called with a `Lit` as a `srcOffset`. So it's more general, but cannot simplify some of the things that the more specific one can. It definitely helps in some cases, as per Zoe's issue

Note that the SMT overflow vs how the EVM actually works is slightly different on CopySlice. We need this rewrite rule, but it means that in some cases, the SMT will differ from the simplification system, when writing to extremely large offsets -- which always leads to over-gas-limit, so it's not a relevant case. However, the rewrite rule actually triggers, in real-world scenarios, and is useful.

NOTE: this is a *cherry-pick from Zoe's changeset* so it will apply perfectly when this is merged, appearing as less of a changeset for Zoe's PR.

## Checklist

- [x] tested locally
- [ ] added automated tests
- [ ] updated the docs
- [x] updated the changelog

### CODE REVIEW

File: test/test.hs [hunk 1]
```diff
@@ -515,6 +515,21 @@ tests = testGroup "hevm"
...
         b <- checkEquiv e (Expr.simplify e)
         assertBoolM "Simplifier failed" b
+    , test "simp-readByte" $ do
+      let srcOffset = (ReadWord (Lit 0x1) (AbstractBuf "stuff1"))
+          size =(ReadWord (Lit 0x1) (AbstractBuf "stuff2"))
+          src = (AbstractBuf "stuff2")
+          e = ReadByte (Lit 0x0) (CopySlice srcOffset (Lit 0x10) size src (AbstractBuf "dst"))
+          simp =Expr.simplify e
+      assertEqualM "readByte simplification" simp (ReadByte (Lit 0x0) (AbstractBuf "dst"))
+    , test "simp-readByte" $ do
```
Comment (blishko) [inline test/test.hs:525]: Should this test use a different name than the previous one?

File: test/test.hs [hunk 2]
```diff
@@ -515,6 +515,21 @@ tests = testGroup "hevm"
...
+      let srcOffset = (ReadWord (Lit 0x1) (AbstractBuf "stuff1"))
+          size =(Lit 0x1)
+          src = (AbstractBuf "stuff2")
+          e = ReadByte (Lit 0x0) (CopySlice srcOffset (Lit 0x10) size src (AbstractBuf "dst"))
+          simp =Expr.simplify e
+      res <- checkEquiv e simp
+      assertEqualM "max-buflength rules"  res True
```
Comment (blishko) [inline test/test.hs:532]: The message is should something else, no?

File: test/test.hs [hunk 3]
```diff
@@ -515,6 +515,21 @@ tests = testGroup "hevm"
...
+          e = ReadByte (Lit 0x0) (CopySlice srcOffset (Lit 0x10) size src (AbstractBuf "dst"))
+          simp =Expr.simplify e
+      assertEqualM "readByte simplification" simp (ReadByte (Lit 0x0) (AbstractBuf "dst"))
+    , test "simp-readByte" $ do
+      let srcOffset = (ReadWord (Lit 0x1) (AbstractBuf "stuff1"))
+          size =(Lit 0x1)
+          src = (AbstractBuf "stuff2")
+          e = ReadByte (Lit 0x0) (CopySlice srcOffset (Lit 0x10) size src (AbstractBuf "dst"))
+          simp =Expr.simplify e
+      res <- checkEquiv e simp
```
Comment (blishko) [inline test/test.hs:531]: Wouldn't it be stronger to assert what the result of the simplification is?
I don't quite understand the purpose of this second test.

File: test/test.hs [hunk 4]
```diff
@@ -562,7 +577,9 @@ tests = testGroup "hevm"
         a = BufLength (ConcreteBuf "ab")
         simp = Expr.simplify a
       assertEqualM "Must be simplified down to a Lit" simp (Lit 2)
-    , test "CopySlice-overflow" $ do
+    -- This test no longer works or needed, due to ReadWord simplification. Notice that it's writing
```
Comment (blishko) [inline test/test.hs:580]: Wait, so does this fail now or is it simply no longer needed?
I think the original purpose of the test was to check this weird overflow behaviour when the copy slice actually wraps back to start reading from the beginning again.
But that is such a can of worms and the SMT semantics seems to be different than EVM semantics.

Maybe we should just remove the test completely?

File: test/test.hs [hunk 5]
```diff
@@ -515,6 +515,21 @@ tests = testGroup "hevm"
         let e = BufLength (CopySlice (Lit 0x2) (Lit 0x2) (Lit 0x1) (ConcreteBuf "") (ConcreteBuf ""))
         b <- checkEquiv e (Expr.simplify e)
         assertBoolM "Simplifier failed" b
+    , test "simp-readByte" $ do
+      let srcOffset = (ReadWord (Lit 0x1) (AbstractBuf "stuff1"))
+          size =(ReadWord (Lit 0x1) (AbstractBuf "stuff2"))
+          src = (AbstractBuf "stuff2")
+          e = ReadByte (Lit 0x0) (CopySlice srcOffset (Lit 0x10) size src (AbstractBuf "dst"))
+          simp =Expr.simplify e
```
Comment (blishko) [inline test/test.hs:523]:
```suggestion
          simp = Expr.simplify e
```

File: test/test.hs [hunk 6]
```diff
@@ -515,6 +515,21 @@ tests = testGroup "hevm"
         let e = BufLength (CopySlice (Lit 0x2) (Lit 0x2) (Lit 0x1) (ConcreteBuf "") (ConcreteBuf ""))
         b <- checkEquiv e (Expr.simplify e)
         assertBoolM "Simplifier failed" b
+    , test "simp-readByte" $ do
+      let srcOffset = (ReadWord (Lit 0x1) (AbstractBuf "stuff1"))
+          size =(ReadWord (Lit 0x1) (AbstractBuf "stuff2"))
```
Comment (blishko) [inline test/test.hs:520]:
```suggestion
          size = (ReadWord (Lit 0x1) (AbstractBuf "stuff2"))
```

**Review by blishko — COMMENTED** [review]
The simplification is sound, but only if writing to a buffer does not overflow.
In that case we have a conflict between the SMT semantics and EVM semantics.
I guess we have agreed to ignore this weird edge case, right?

**Comment by blishko** [comment]
Also, there seem to be some test failures on windows:
```
FAIL
      Exception: Maybe.fromJust: Nothing
      CallStack (from HasCallStack):
        error, called at libraries\base\Data\Maybe.hs:150:21 in base:Data.Maybe
        fromJust, called at test\EVM\Test\Tracing.hs:880:30 in hevm-0.54.2-inplace-test:EVM.Test.Tracing
      Use -p '/calldata-wraparound-1/' to rerun this test only.
    calldata-wraparound-2:                                        Error converting trace! exit code:ExitFailure 127
```

File: test/test.hs (hunk 3 above)
Comment (msooseth) [reply test/test.hs:531]: I guess I wanted to prove that the simplification and the SMT solver agree... maybe I was a bit too eager to add 2 tests. It does prove that the SMT solver agrees, which is nice. Let me know what you think. Maybe it can be removed, honestly.

File: test/test.hs (hunk 4 above)
Comment (msooseth) [reply test/test.hs:580]: Good point! Yeah, removing!

**Comment by msooseth** [comment]
The windows build issue is unrelated, unfortunately. It's occurring here too:

https://github.com/ethereum/hevm/actions/runs/13283178489/job/37085893025

which is also unrelated. I'll merge this with that issue, and I have already asked the Echidna folks to look into the Windows issue. I think it's something small, or at least I'm hoping :) Probably some shell script thing.

**Comment by msooseth** [comment]
> The simplification is sound, but only if writing to a buffer does not overflow. In that case we have a conflict between the SMT semantics and EVM semantics. I guess we have agreed to ignore this weird edge case, right?

Yep, we did :)

File: test/test.hs (hunk 3 above)
Comment (blishko) [reply test/test.hs:531]: Right, I see. I guess it is good to have a test for that as well!

Comment (blishko) [reply test/test.hs:531]: So, keep the test!
But please fix the spaces around `=` also in this test, as you've already done for the first one :)

Comment (msooseth) [reply test/test.hs:531]: Oh man, I'm so sorry. I reviewed this PR and then I was like... let me add some tests. And then of course all the mess is from the tests... Sorry.

**Comment by msooseth** [comment]
@blishko beware -- I added the same simplification to `readWord` as well, which I now noticed is missing.

File: test/test.hs [hunk 7]
```diff
@@ -515,6 +515,21 @@ tests = testGroup "hevm"
...
+    , test "simp-readByte1" $ do
+      let srcOffset = (ReadWord (Lit 0x1) (AbstractBuf "stuff1"))
+          size = (ReadWord (Lit 0x1) (AbstractBuf "stuff2"))
+          src = (AbstractBuf "stuff2")
+          e = ReadByte (Lit 0x0) (CopySlice srcOffset (Lit 0x10) size src (AbstractBuf "dst"))
+          simp = Expr.simplify e
+      assertEqualM "readByte simplification" simp (ReadByte (Lit 0x0) (AbstractBuf "dst"))
+    , test "simp-readByte2" $ do
+      let srcOffset = (ReadWord (Lit 0x1) (AbstractBuf "stuff1"))
+          size =(Lit 0x1)
```
Comment (blishko) [inline test/test.hs:527]:
```suggestion
          size = (Lit 0x1)
```

File: test/test.hs [hunk 8]
```diff
@@ -515,6 +515,21 @@ tests = testGroup "hevm"
...
+          src = (AbstractBuf "stuff2")
+          e = ReadByte (Lit 0x0) (CopySlice srcOffset (Lit 0x10) size src (AbstractBuf "dst"))
+          simp =Expr.simplify e
```
Comment (blishko) [inline test/test.hs:530]:
```suggestion
          simp = Expr.simplify e
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -19,6 +19,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Multiple solutions are allowed for a single symbolic expression
 - Aliasing works much better for symbolic and concrete addresses
 - Constant propagation for symbolic values
+- One more simplification rule for `ReadByte` when the `CopySlice` after it
```
Comment (blishko) [inline CHANGELOG.md:22]: Please mention also the added simplification for `ReadWord` in the entry.

**Review by blishko — COMMENTED** [review]
Maybe you should also add a test for the `ReadWord` simplification?

File: CHANGELOG.md (hunk 9 above)
Comment (msooseth) [reply CHANGELOG.md]: Oh, good catch. Fixed!

**Comment by msooseth** [comment]
> Maybe you should also add a test for the `ReadWord` simplification?

Yes.. sorry. I just added two :)

**Review by blishko — APPROVED** [review]
LGTM!
